### PR TITLE
feat: add audit trail for channel updates

### PR DESCRIPTION
**Pull request by vishnu-narayanan**
# Pull Request Template

## Description

Fixes https://linear.app/chatwoot/issue/CW-2027/subsequent-update-events-to-inbox-object-are-not-logged

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: enterprise/app/models/enterprise/channelable.rb [hunk 1]
```diff
@@ -1,22 +1,31 @@
...
-  def create_audit_log_entry
-    account = self.account
-    associated_type = 'Account'
-    auditable_id = inbox.id
-    auditable_type = 'Inbox'
-    audited_changes = saved_changes.except('updated_at')
+  # Active support concern has `included` which changes the order of the method lookup chain
+  # https://stackoverflow.com/q/40061982/3824876
+  # manually prepend the instance methods to combact this
+  included do
```
Comment (sojan-official) [inline enterprise/app/models/enterprise/channelable.rb:7]: 🔥

**Review by sojan-official — COMMENTED** [review]
lets also add an ee spec to cover the case of channel update and that appropriate log is created

**Comment by vishnu-narayanan** [comment]
added specs

File: spec/enterprise/models/inbox_spec.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@
 require 'rails_helper'
 
 RSpec.describe Inbox do
+  # let!(:channel) { create(:channel_widget) }
```
[the remark this answers is not in the document]
Comment (vishnu-narayanan) [reply spec/enterprise/models/inbox_spec.rb:6]: done

File: spec/enterprise/models/inbox_spec.rb [hunk 3]
```diff
@@ -40,14 +41,31 @@
...
+        inbox.channel.update(widget_color: new_color)
+
+        # check if channel update creates an audit log against inbox
+        expect(Audited::Audit.where(auditable_type: 'Inbox', action: 'update').count).to eq(2)
+        # Check for the specific widget_color update in the audit log
+        expect(Audited::Audit.where(auditable_type: 'Inbox', action: 'update',
+                                    audited_changes: { 'widget_color' => [previous_color, new_color] }).count).to eq(1)
+      end
+    end
+
```
[the remark this answers is not in the document]
Comment (vishnu-narayanan) [reply spec/enterprise/models/inbox_spec.rb]: done

**Comment by vishnu-narayanan** [comment]
Refactored as per review feedback
- Move channel update changes to separate context block
- Add a separate channel (API) spec for audits

**Comment by sojan-official** [comment]
note: @vishnu-narayanan  in case of the failing specs, it seems like the inbox is nil 
```
 undefined method `id' for nil:NilClass
     # ./enterprise/app/models/enterprise/channelable.rb:15:in `create_audit_log_entry'
```

please check whether we have inbox object created in association with these channels as well in the specs.

File: enterprise/app/models/enterprise/channelable.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  module InstanceMethods
+    def create_audit_log_entry
+      account = self.account
+      associated_type = 'Account'
+      auditable_id = inbox.id
+      auditable_type = 'Inbox'
+      audited_changes = saved_changes.except('updated_at')
+
+      return if audited_changes.blank?
```
Comment (sojan-official) [inline enterprise/app/models/enterprise/channelable.rb:19]: Could also add a condition return if inbox.nil? 

for the cases where we don't want this to be triggered , ie inbox delete

Comment (vishnu-narayanan) [reply enterprise/app/models/enterprise/channelable.rb:22]: done

**Review by sojan-official — APPROVED** [review]

**Pull request merged**